vortex-io/src/object_store.rs

```diff
@@ -92,10 +93,22 @@ impl VortexReadAt for ObjectStoreReadAt {
                 )
                 .await?;
 
-            let mut byte_stream = response.into_stream();
-            while let Some(bytes) = byte_stream.next().await {
-                let bytes = bytes?;
-                buf.extend_from_slice(&bytes);
+            // NOTE: ObjectStore specializes the payload based on if it is backed by a File or if
+            //  it's coming from a network stream. Internally they optimize the File implementation
+            //  to only perform a single allocation when calling `.bytes().await`, which we
+            //  replicate here by emitting the contents directly into our aligned buffer.
+            match response.payload {
+                GetResultPayload::File(file, _) => {
+                    unsafe {
+                        buf.set_len(len as _);
+                    }
+                    file.read_exact_at(&mut buf, pos)?;
+                }
+                GetResultPayload::Stream(mut byte_stream) => {
+                    while let Some(bytes) = byte_stream.next().await {
+                        buf.extend_from_slice(&bytes?);
```

Comment:
ok read the code, we should do it ourselves IMO (set the length and capacity and copy the data).

Reply:
in the case of HTTP2 it will be one Bytes per DATA frame, in the http1 instance I'm not sure what it is but either way we can't pass in a buffer to the HTTP client, afaict

Reply:
as far as I can tell it copies one **byte** at a time, not `Bytes`, give me a sec i'll have a gist/some numbers to see if I'm saying anything valuable.

Reply:
Capturing slack convo: it appears that Vec::extend_from_slice(&Bytes) will go through the pathway that uses a copy_nonoverlapping under the hood

![image](https://github.com/user-attachments/assets/4e334d3d-263f-4a60-8bd6-0abc4803d559)